distributed/shuffle/_core.py

```diff
@@ -0,0 +1,278 @@
+
+    async def _flush_comm(self) -> None:
+        self.raise_if_closed()
+        await self._comm_buffer.flush()
+
+    async def flush_receive(self) -> None:
+        self.raise_if_closed()
+        await self._disk_buffer.flush()
+
+    async def close(self) -> None:
+        if self.closed:  # pragma: no cover
+            await self._closed_event.wait()
+            return
+
+        self.closed = True
+        await self._comm_buffer.close()
+        await self._disk_buffer.close()
+        self._closed_event.set()
+
+    def fail(self, exception: Exception) -> None:
+        if not self.closed:
+            self._exception = exception
+
+    def _read_from_disk(self, id: NDIndex) -> bytes:
+        self.raise_if_closed()
+        data: bytes = self._disk_buffer.read("_".join(str(i) for i in id))
+        return data
+
+    async def receive(self, data: list[tuple[_T_partition_id, bytes]]) -> None:
+        await self._receive(data)
+
+    async def _ensure_output_worker(self, i: _T_partition_id, key: str) -> None:
+        assigned_worker = self._get_assigned_worker(i)
+
+        if assigned_worker != self.local_address:
+            result = await self.scheduler.shuffle_restrict_task(
+                id=self.id, run_id=self.run_id, key=key, worker=assigned_worker
+            )
+            if result["status"] == "error":
+                raise RuntimeError(result["message"])
+            assert result["status"] == "OK"
+            raise Reschedule()
+
+    @abc.abstractmethod
+    def _get_assigned_worker(self, i: _T_partition_id) -> str:
+        """Get the address of the worker assigned to the output partition"""
+
+    @abc.abstractmethod
+    async def _receive(self, data: list[tuple[_T_partition_id, bytes]]) -> None:
+        """Receive shards belonging to output partitions of this shuffle run"""
+
+    @abc.abstractmethod
+    async def add_partition(
+        self, data: _T_partition_type, partition_id: _T_partition_id
+    ) -> int:
+        """Add an input partition to the shuffle run"""
+
+    @abc.abstractmethod
+    async def get_output_partition(
+        self, partition_id: _T_partition_id, key: str, meta: pd.DataFrame | None = None
+    ) -> _T_partition_type:
+        """Get an output partition to the shuffle run"""
+
+
+def get_worker_plugin() -> ShuffleWorkerPlugin:
+    from distributed import get_worker
+
+    try:
+        worker = get_worker()
+    except ValueError as e:
+        raise RuntimeError(
+            "`shuffle='p2p'` requires Dask's distributed scheduler. This task is not running on a Worker; "
+            "please confirm that you've created a distributed Client and are submitting this computation through it."
+        ) from e
+    plugin: ShuffleWorkerPlugin | None = worker.plugins.get("shuffle")  # type: ignore
+    if plugin is None:
+        raise RuntimeError(
+            f"The worker {worker.address} does not have a ShuffleExtension. "
+            "Is pandas installed on the worker?"
```

Comment:
Is that now necessary even if only using shuffle extensions for array rechunking?

Reply:
This exception is outdated, but the worker plugin is still involved in array rechunking. Note that we have conflicting names between the generic "P2P shuffle" approach as in sending data all-to-all across the cluster and the specific `dataframe.shuffle` operation. If you have naming suggestions for resolving this, please let me know. As with other comments, I would rather not address this here.